File: frontend/src/api/kanban.ts
import {
  CreatePost as APICreatePost,
  DeleteBoard as APIDeleteBoard,
  DeletePost as APIDeletePost,
  UpdatePost as APIUpdatePost,
  CreateKanban,
  GetKanbanBoardsFromUser,
  GetKanbanBord,
  UpdateKanban,
} from "@bindings/viktor/backend/app";
import type { Kanban } from "@bindings/viktor/db/models";
import { z } from "zod";
import { CheckSession } from "./userdata";

export const CreateBoard = async (BoardName: string): Promise<boolean> => {
  const session = await CheckSession();
  if (session == null) return false;
  if (session.Id == null) return false;
  if (session.UserId == null) return false;
  return CreateKanban(session.UserId, BoardName);
};

export const PostParams = z.object({
  Name: z.string(),
  Description: z.string().optional(),
  Importance: z.string(),
  Status: z.string(),
});

type PostParams = z.infer<typeof PostParams>;

export const CreatePost = async (
  BoardId: string,
  Params: PostParams
): Promise<boolean> => {
  return await APICreatePost(
    Params.Name,
    Params.Status,
    Params.Importance,
    Params.Description ? Params.Description : null,
    BoardId
  );
};

export const GetBoardsFromUser = async (): Promise<Array<Kanban> | null> => {
  const session = await CheckSession();
  if (session == null) return null;
  if (session.Id == null) return null;
  if (session.UserId == null) return null;
  return await GetKanbanBoardsFromUser(session.UserId);
};

export const GetBoard = async (BoardId: string): Promise<Kanban | null> => {
  return await GetKanbanBord(BoardId);
};

export const UpdateBoard = async (
  BoardId: string,
  Name: string
): Promise<boolean> => {
  return await UpdateKanban(BoardId, Name);
};

export const UpdatePost = async (
  PostId: string,
  Params: PostParams
): Promise<boolean> => {
  return await APIUpdatePost(
    PostId,
    Params.Name,
    Params.Status,
    Params.Importance,
    Params.Description ? Params.Description : null
  );
};

export const DeletePost = async (PostId: string): Promise<boolean> => {
  return await APIDeletePost(PostId);
};

export const DeleteBoard = async (BoardId: string): Promise<boolean> => {
  return await APIDeleteBoard(BoardId);
};

File: backend/kanban.go
package backend

import (
	"fmt"
	"viktor/db"

	"github.com/wailsapp/wails/v3/pkg/application"
)

func (a *App) CreateKanban(id string, Name string) bool {
	_, err := a.DB.CreateKanban(Name, id)
	if err != nil {
		dialog := application.ErrorDialog()
		dialog.SetTitle("FEHLER!")
		dialog.SetMessage(fmt.Sprintf("%s: %s", "[CreateKanban] Fehler beim Anlegen", err))
		dialog.Show()
		return false
	}
	return true
}

func (a *App) CreatePost(Name, Status, Importance string, Description *string, KId string) bool {
	_, err := a.DB.CreatePost(KId, Name, Description, Status, Importance)
	if err != nil {
		dialog := application.ErrorDialog()
		dialog.SetTitle("FEHLER!")
		dialog.SetMessage(fmt.Sprintf("%s: %s", "[CreatePost] Fehler beim Anlegen", err))
		dialog.Show()
		return false
	}
	return true
}

func (a *App) GetKanbanBord(id string) *db.Kanban {
	board, err := a.DB.GetKanban(id)
	if err != nil {
		dialog := application.ErrorDialog()
		dialog.SetTitle("FEHLER!")
		dialog.SetMessage(fmt.Sprintf("%s: %s", "[GetKanbanBord] Fehler", err))
		dialog.Show()
		return nil
	}
	return board
}

func (a *App) GetKanbanBoardsFromUser(id string) []db.Kanban {
	boards, err := a.DB.GetAllKanbans(id)
	if err != nil {
		dialog := application.ErrorDialog()
		dialog.SetTitle("FEHLER!")
		dialog.SetMessage(fmt.Sprintf("%s: %s", "[GetKanbanBoardsFromUser] Fehler", err))
		dialog.Show()
		return nil
	}
	return boards
}

func (a *App) GetPostsFromBoard(id string) []db.Post {
	posts, err := a.DB.GetAllPostFromBoard(id)
	if err != nil {
		dialog := application.ErrorDialog()
		dialog.SetTitle("FEHLER!")
		dialog.SetMessage(fmt.Sprintf("%s: %s", "[GetPostsFromBoard] Fehler", err))
		dialog.Show()
		return nil
	}
	return posts
}

func (a *App) UpdateKanban(id, Name string) bool {
	err := a.DB.UpdateKanban(id, Name)
	if err != nil {
		dialog := application.ErrorDialog()
		dialog.SetTitle("FEHLER!")
		dialog.SetMessage(fmt.Sprintf("%s: %s", "[UpdateKanban] Fehler", err))
		dialog.Show()
		return false
	}
	return true
}

func (a *App) UpdatePost(id, Name, Status, Importance string, Description *string) bool {
	err := a.DB.UpdatePost(id, Name, Description, Status, Importance)
	if err != nil {
		dialog := application.ErrorDialog()
		dialog.SetTitle("FEHLER!")
		dialog.SetMessage(fmt.Sprintf("%s: %s", "[UpdatePost] Fehler beim Speichern", err))
		dialog.Show()
		return false
	}
	return true
}

func (a *App) DeletePost(id string) bool {
	err := a.DB.DeletePost(id)
	if err != nil {
		dialog := application.ErrorDialog()
		dialog.SetTitle("FEHLER!")
		dialog.SetMessage(fmt.Sprintf("%s: %s", "[DeletePost] Fehler", err))
		dialog.Show()
		return false
	}
	return true
}

func (a *App) DeleteBoard(id string) bool {
	err := a.DB.DeleteKanban(id)
	if err != nil {
		dialog := application.ErrorDialog()
		dialog.SetTitle("FEHLER!")
		dialog.SetMessage(fmt.Sprintf("%s: %s", "[DeleteBoard] Fehler", err))
		dialog.Show()
		return false
	}
	return true
}
